silkworm/common/log.hpp

```diff
@@ -27,31 +28,40 @@ enum LogLevels {
     LogTrace, LogDebug, LogInfo, LogWarn, LogError, LogCrit, LogNone
 };
 
-// Log to two output streams - typically the console and a log file.
+// Log to one or two output streams - typically the console and optional log file.
 class Logger {
 public:
-    Logger(std::ostream& o1, std::ostream& o2, LogLevels level=LogNone)
+    Logger(LogLevels level=LogInfo, std::ostream& o1=std::cerr, std::ostream& o2=Logger::null_stream())
     : stream(o1, o2), verbosity(level) {}
 
     int level(LogLevels level) { return verbosity = level; }
     int level() { return verbosity; }
 
-    std::ostream& log(LogLevels level, const char* file, int line){
-    std::time_t time = std::time(nullptr);
-    return stream << LogTags[level] \
-                  << std::put_time(std::gmtime(&time), " %Y-%m-%d %H:%M:%S %Z ") \
-                  << file << ":" << line << "| ";
+    std::ostream& log(LogLevels level, const char* file, int line) {
+        std::time_t time = std::time(nullptr);
+        return stream << LogTags[level]
+                      << std::put_time(std::gmtime(&time), " %Y-%m-%d %H:%M:%S %Z ")
+                      << file << ":" << line << "| ";
     }
+
+    static Logger& default_logger();
+    static std::ostream& null_stream();
+
 private:
     teestream stream;
     LogLevels verbosity;
-    static constexpr char const LogTags[7][6] = { "TRACE", "DEBUG", "INFO ", "WARN ", "ERROR", "CRIT ", "NONE" };
+    static constexpr char const LogTags[7][6] = {
+        "TRACE", "DEBUG", "INFO ", "WARN ", "ERROR", "CRIT ", "NONE"
+    };
 };
 
-// log to whatever logger is in scope
-#define SILKWORM_LOG(level_) \
-    if (logger.level() < (level_)) {} \
-    else logger.log(level_, __FILE__, __LINE__)
+#define SILKWORM_LOG_TO(logger_, level_) \
+    if ((logger_).level() < (level_)) {} \
+    else (logger_).log((level_), __FILE__, __LINE__)
+#define SILKWORM_LOG_TO_LEVEL(logger_, level_) ((logger_).level(level_))
```

Comment:
There's no need for it to be a macro, is there? We should [avoid macros if possible](https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros).

Reply:
Yeah - turns out 'commit -a' doesn't do it

It needs to be a macro to avoid evaluating the `log(level,file,line)` call and the subsequent `<<` ops when the logging level is too low.

Reply:
OK, but SILKWORM_LOG_TO_LEVEL doesn't have to be a macro, does it?

Reply:
I find it much more clear to have four consistently named logging macros than two macros and two raw C++ calls.  It also makes using and changing the code behind the macros easier -- who wants to remember that the default logger is currently accessed by reference via a static member function of a class rather than one of the other ways it might be?